### DISCUSSION TITLE
Add mappers to cursors

### DISCUSSION THREAD

**Discussion by alfonsocv12** [sqlalchemy/sqlalchemy#9386] [Usage Questions]
### Describe the use case

I'm creating a variable query_builder, where you can filter out your columns, currently I'm usign select and Tables() for this.

I use a hexagonal architecture so I have a domain entity

```python
@dataclass
class Foo:
   id: Optional[UUID]
   name: Optional[str]
   foo: Optional[str]
```

```python
foo = Table(
    'foo',
    metadata_obj,
    Column('id', string, primary_key=True),
    Column('name', String),
    Column('foo', String),
)
```

```python
fields = ('name', 'foo')
select(table.c[fields])
```

Currently I'm overriding the all function of the execute output this way

```python
mapper = create_mapper_class(fields)

def all(self)  -> List[_T]:
    post_creational_filter = self._post_creational_filter

    make_row = self._row_getter

    rows = self._fetchall_impl()
    made_rows: List[_InterimRowType[_R]]
    if make_row:
        ### This is the only thing changing
        made_rows = [mapper(*row) for row in rows]
    else:
        made_rows = rows  # type: ignore

    interim_rows: List[_R]

    if self._unique_filter_state:
        uniques, strategy = self._unique_strategy

        interim_rows = [
            made_row  # type: ignore
            for made_row, sig_row in [
                (
                    made_row,
                    strategy(made_row) if strategy else made_row,
                )
                for made_row in made_rows
            ]
            if sig_row not in uniques and not uniques.add(sig_row)  # type: ignore # noqa: E501
        ]
    else:
        interim_rows = made_rows  # type: ignore

    if post_creational_filter:
        interim_rows = [
            post_creational_filter(row) for row in interim_rows
        ]
    return interim_rows

cursor = session.execute(query_builder(Foo, ("name", "foo")))
cursor.all = MethodType(all, cursor)

cursor.all()
```

### Databases / Backends / Drivers targeted

postgrest

### Example Use

I will like to add the functionality of mappers in cursors

```python
select(mapper, table.c[fields_tuple])
```

this way I have a custom mapper for every fetch instead of needing to overwrite everything

### Additional context

I believe this would be a great optimization on big queries.

I can start building the feature, I'm pretty familiar with `Result class` at the moment.

But I will like to know if somebody has a better approach to the problem

**Comment by CaselIT** [comment]
Hi,

You seem to be reinvented the orm.
Can't you just map Foo to the table and query it using the orm?
What is the advantage of this compared to the orm?

**Reply by alfonsocv12** [reply]
Okay so your suggestion is to add Foo to the registry imperatively with the table.

The problem with this is even if I add a select `.from_statement(select(*my_rows))` I will get a select (all_attributes from Foo).

And if I try to add multiple classes to the same table I will need to make a remove, add mess on the registry

**Reply by CaselIT** [reply]
You can use the load only option to control what column to load https://docs.sqlalchemy.org/en/20/orm/queryguide/columns.html#sqlalchemy.orm.load_only

If the list of columns is static, you can also use deferred statically (and override only if needed) https://docs.sqlalchemy.org/en/20/orm/queryguide/columns.html#orm-queryguide-deferred-imperative

**Reply by alfonsocv12** [reply]
Let me try that one thanks

**Reply by alfonsocv12** [reply]
This still brings everything is the same as a select *

```python
query = (
    select(Comment)
        .options(load_only(Comment.id, Comment.uuid))
        .where(Comment.object_id == 1)
        .limit(10)
        .offset(0)
)
print(session.scalars(query).all())
```

Output
```
[Comment(user_id=Foo, content='Foo, object_id=Foo, id=Foo, uuid=UUID('foo'), created_at=datetime.datetime(foo, tzinfo=foo), updated_at=datetime.datetime(foo, tzinfo=foo), likes=foo, answers=foo),
```

I know is weird using id and uuid. I'm not sure what this decision was made.

Table and class
```python
@dataclass
class Comment:
    user_id: Optional[int]
    content: Optional[str]
    object_id: Optional[int]
    id: Optional[int] = None  # internal id, require for retro compatibility
    uuid: Optional[UUID] = field(default_factory=uuid4)
    created_at: Optional[datetime] = field(default_factory=datetime.now)
    updated_at: Optional[datetime] = field(default_factory=datetime.now)
    likes: Optional[int] = 0
    answers: Optional[int] = 0
  
comment_table = Table(
    'comment',
    metadata_obj,
    Column('id', Integer, primary_key=True),
    Column('content', String),
    Column('uuid', String),
    Column('object_id', Integer),
    Column('user_id', Integer),
    Column('created_at', DateTime),
    Column('updated_at', DateTime),
)

mapper_registry.map_imperatively(Comment, comment_table)
```

**Reply by CaselIT** [reply]
that's the repr that accesses the arguments and make sqlalchemy lazy load. look into the class `__dict__` or replace the repr with one that only prints what's present in the dict.

You can confirm this by using `load_only(Comment.id, Comment.uuid, raiseload=True)`.

Note that if you plan on using dataclasses, you can't use the default repr, since it will break if you don't have all the attributes loaded.

**Comment by zzzeek** [comment]
> I believe this would be a great optimization on big queries.

an optimization problem should be approached by looking at the current use case and doing profiling.   Can you describe the current performance problem using SQLAlchemy's provided APIs ?

**Reply by alfonsocv12** [reply]
Sure let me add some numbers to make my case stronger on why this optimizes anything
